dev/scripts/build.ts

```diff
@@ -3,46 +3,139 @@ import { join } from "node:path"
 import { readJson, writeJson } from "../attest/src/fs.js"
 import { shell } from "../attest/src/shell.js"
 import { repoDirs } from "./common.js"
+import { rewritePaths } from "./overwrite.js"
+import type { ReplacementDictionary } from "./overwrite.js"
 
 const packageRoot = process.cwd()
 const outRoot = join(packageRoot, "dist")
 const packageJson = readJson(join(packageRoot, "package.json"))
+const tempTsConfigBaseName = "tsconfig.temp"
+const tempTsConfigPath = join(packageRoot, `${tempTsConfigBaseName}.json`)
 
-const buildFormat = (module: "commonjs" | "esnext") => {
-    const outDir = join(outRoot, module === "commonjs" ? "cjs" : "mjs")
-    writeJson(tempTsConfig, {
-        ...tsConfigData,
-        include: ["src"],
+const nuke = (target: string) =>
+    rmSync(target, { recursive: true, force: true })
+const clone = (from: string, to: string): void =>
+    cpSync(from, to, { recursive: true, force: true })
+
+const writeManifest =
+    (overrides: Record<string, unknown>) =>
+    (sourceDir: string, targetDir: string) => {
+        const manifest = readJson(join(sourceDir, "package.json"))
+        writeJson(join(targetDir, "package.json"), {
+            ...manifest,
+            ...overrides
+        })
+    }
+
+const Sources = {
+    utils: ["dev", "utils"],
+    attest: ["dev", "attest"]
+} as const
+
+const replacementDictionary: ReplacementDictionary = {
+    attest: {
+        pattern: /"(\.\.\/)+[^"]*attest\/[^"]*\.js"/g,
+        replacement: `"@arktype/attest"`
+    },
+    utils: {
+        pattern: /"(\.\.\/)+[^"]*utils\/[^"]*\.js"/g,
```

Comment:
Calling this out: these are hardcoded. Ideally we'd derive these using the `RegExp` constructor, and that's how I originally had things set up, but I ran into trouble getting it to work.

I was using a helper that looked something like this:

```typescript
const makePattern = (dir: string) => 
new RegExp('"(\\.\\.\\/)+[^"]*' + dir + '\\/[^"]*\\.js"', 'g')
```

^^ that seemed to work just fine until actually added to the build.

Regarding the regex itself, this one should be pretty performant. It uses a negative set (`[^"]*`) to minimize backtracking, which cut down the number of iterations the engine needed to make over a single line from around 200 to less than 50. I believe this gets it down to linear time complexity too, although I'm not 100% on that.

There are probably additional optimizations I could make here, but didn't want to spend too much time here.

Reply:
That's great! I think you had the right instinct with the helper method, but this is an acceptable amount and context for this sort of repetition.

Our local build is also not a particularly perf-sensitive context, especially given how far we've gone out of our way to avoid having to do it other than to publish, so while the optimizations are certainly appreciated, don't feel obligated to consider the perf implications in such depth outside the core package.

Reply:
It ended up being necessary in this case.

My first attempt looked something like this:

`  pattern: /"(\.\.\/)+(.+)+(utils)\/.+\.js"/g,`

If you're really curious you can try it and see how long the build takes.

This is the only version I came up with that didn't add 20+ seconds to the build.